portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -90,23 +91,40 @@ private String convertToJsonUsingGson(List<Comment> serverStats) {
 @Override
 public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
-    String comment = request.getParameter("comment-input");
-    String name = request.getParameter("name-input");
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/index.html");
+      return;
+    }
 
-    Comment newComment = new Comment(name,comment);
-   // String json = convertToJsonUsingGson(newComment);
-   // comments.add(newComment);
+    String comment = request.getParameter("comment-input");
+    String name =  getUserNickname(userService.getCurrentUser().getUserId());
 
     response.setContentType("text/html;");
 
     Entity taskEntity = new Entity("Comment");
     taskEntity.setProperty("comment", comment);
     taskEntity.setProperty("name", name);
+    taskEntity.setProperty("timestamp", System.currentTimeMillis());
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(taskEntity);
 
     response.sendRedirect("/index.html");
   }
-}
 
+   private String getUserNickname(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("UserInfo")
+            .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return null;
```

Comment:
FYI, methods that only sometimes return a result are pretty common -- and a frequent source of bugs.

Java 8 has introduced a new set of utilities for dealing with this: `Optional<String>`. You'll see this used in many modern Java libraries.

Null checks are still pretty common too, but knowing about `Optional<>` types is useful if you work more with Java.